File: 2020/src/aoc2020/Day4.java
package aoc2020;

import java.io.*;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Day4 {
	private ArrayList<String> input = new ArrayList<>();
	
	Day4() throws FileNotFoundException {
		Scanner scan = new Scanner(new File("data/data4.txt"));
		String k = "";
		while (scan.hasNextLine()) {
			String s = scan.nextLine();
			if (s.isEmpty()) {
				input.add(k);
				k = "";
			} else {
				k += s + " ";
			}
		}
		if (!k.isEmpty()) {
			input.add(k);
		}
	}
	
	private void part1() {
		String[] required = {"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"};
		int res = 0;
		for (String passport : input) {
			boolean good = true;
			for (String key: required) {
				if (!passport.contains(key + ":")) {
					good = false;
					break;
				}
			}
			if (good) res++;
		}
		System.out.println(res);
	}
	
	private void part2() {
		int res = 0;
		Pattern[] patterns = {
			Pattern.compile("byr:(19[2-9]\\d|200[0-2])"),
			Pattern.compile("iyr:(201\\d|2020)"),
			Pattern.compile("eyr:(202\\d|2030)"),
			Pattern.compile("hgt:(((59|6\\d|7[0-6])in)|(1([5-8]\\d|9[0-3])cm))"),
			Pattern.compile("hcl:(#[0-9a-f]{6})"),
			Pattern.compile("ecl:(amb|blu|brn|gry|grn|hzl|oth)"),
			Pattern.compile("pid:\\d{9}[^\\d]"),
		};
		for (String passport : input) {
			boolean good = true;
			for (Pattern pattern: patterns) {
				Matcher m = pattern.matcher(passport);
				if (!m.find()) {
					good = false;
					break;
				}
			}
			if (good) res++;
		}
		System.out.println(res);
	}
	
	public static void main(String[] args) throws Exception {
		Day4 day = new Day4();
		//day.part1();
		day.part2();
	}
	
}

File: 2020/day4.py
#!/usr/bin/pypy3
import re
from utils import is_number

with open("data/data4.txt", "r") as f:
    ps = f.read().split("\n\n")

# Part 1
res = 0
all_fields = ["byr", "iyr", "eyr", "hgt", "ecl", "hcl", "pid"]
for p in ps:
    fs = [f.split(":") for f in p.split()]
    ks = set([k for [k, _] in fs])
    if all([f in ks for f in all_fields]):
        res += 1
print(res)

# Part 2
res = 0
for p in ps:
    fs = {}
    for f in p.split():
        [k, v] = f.split(":")
        fs[k] = v
    if all([f in fs.keys() for f in all_fields]) \
       and is_number(fs["byr"]) and (1920 <= int(fs["byr"]) <= 2002) \
       and is_number(fs["iyr"]) and (2010 <= int(fs["iyr"]) <= 2020) \
       and is_number(fs["eyr"]) and (2020 <= int(fs["eyr"]) <= 2030) \
       and re.match(r"#[A-Fa-f0-9]{6}$", fs["hcl"]) \
       and re.match(r"(amb|blu|brn|gry|grn|hzl|oth)$", fs["ecl"]) \
       and re.match(r"[0-9]{9}$", fs["pid"]) \
       and is_number(fs["hgt"][:-2]) \
       and ((fs["hgt"][-2:] == "cm" and 150 <= int(fs["hgt"][:-2]) <= 193) or
            (fs["hgt"][-2:] == "in" and 59 <= int(fs["hgt"][:-2]) <= 76)):
        res += 1
print(res)
